internal/telemetry/reporter.go

```diff
@@ -0,0 +1,226 @@
+// Package telemetry implements a client for reporting telemetry data used to
+// prioritize development of SpiceDB.
+//
+// For more information, see:
+// https://github.com/authzed/spicedb/blob/main/TELEMETRY.md
+package telemetry
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/gogo/protobuf/proto"
+	"github.com/golang/snappy"
+	dto "github.com/prometheus/client_model/go"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+	"github.com/rs/zerolog/log"
+)
+
+const (
+	DefaultEndpoint = "https://telemetry.authzed.com"
+	MetricPrefix    = "spicedb_telemetry_"
+	Interval        = time.Hour
+)
+
+func writeTimeSeries(ctx context.Context, client *http.Client, endpoint string, ts []*prompb.TimeSeries) error {
+	pbBytes, err := proto.Marshal(&prompb.WriteRequest{
+		Timeseries: ts,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to marshal Prometheus remote write protobuf: %w", err)
+	}
+	compressedPB := snappy.Encode(nil, pbBytes)
+
+	r, err := http.NewRequestWithContext(ctx, http.MethodPost, endpoint, bytes.NewBuffer(compressedPB))
+	if err != nil {
+		return fmt.Errorf("failed to create Prometheus remote write http request: %w", err)
+	}
+
+	r.Header.Add("X-Prometheus-Remote-Write-Version", "0.1.0")
+	r.Header.Add("Content-Encoding", "snappy")
+	r.Header.Set("Content-Type", "application/x-protobuf")
+
+	resp, err := client.Do(r)
+	if err != nil {
+		return fmt.Errorf("failed to send Prometheus remote write: %w", err)
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode/100 != 2 {
+		body, _ := ioutil.ReadAll(resp.Body)
+		return fmt.Errorf(
+			"unexpected Prometheus remote write response: %d: %s",
+			resp.StatusCode,
+			string(body),
+		)
+	}
+
+	return nil
+}
+
+func timestamp(m *dto.Metric) int64 {
+	if ts := m.GetTimestampMs(); ts != 0 {
+		return ts
+	}
+	return time.Now().UnixNano() / int64(time.Millisecond)
+}
+
+func nameLabel(name string) *prompb.Label {
+	return &prompb.Label{
+		Name:  model.MetricNameLabel,
+		Value: name,
+	}
+}
+
+func discoverTimeseries() (allTS []*prompb.TimeSeries, err error) {
+	metricFams, err := registry.Gather()
+	if err != nil {
+		return nil, fmt.Errorf("failed to gather telemetry metrics: %w", err)
+	}
+
+	for _, fam := range metricFams {
+		for _, metric := range fam.GetMetric() {
+			var labels []*prompb.Label
+			for _, label := range metric.GetLabel() {
+				labels = append(labels, &prompb.Label{
+					Name:  label.GetName(),
+					Value: label.GetValue(),
+				})
+			}
+
+			switch *fam.Type {
```

Comment:
I think you can just use https://pkg.go.dev/github.com/prometheus/common/expfmt#ExtractSamples to avoid this, but I didn't test it

Reply:
Looked into it and this looks usable.